---
title: "MSigDF: Molecular Signature Database (MSigDB) in a Data Frame"
author:
- name: Stephen D. Turner 
  affiliation: University of Virginia, Charlottesville VA, United States
  email: vustephen@gmail.com
- name: Enrique M. Toledo 
  affiliation: ---
  email: enriquetoledo@gmail.com
maintenace: Enrique M. Toledo (enriquetoledo@gmail.com)
package: msigdf
output:
  BiocStyle::html_document
abstract: |
  This data package contains the Molecular Signature Database (MSigDB) for both human and predicted mouse orthologs in separate data frames (tibbles). 
  Each data frame (`msigdf.human` and `msigdf.mouse`) contain three columns: the collection (Hallmark, or c1-c8), the gene set, and Entrez IDs
  for genes in that set. The `msigdf.urls` tibble contains links to descriptions on the Broad Institute's website of each gene set.
  **[Source code available on GitHub](https://github.com/ToledoEM/msigdf)**.
vignette: |
  %\VignetteIndexEntry{MSigDF: Molecular Signature Database (MSigDB) in a Data Frame}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
library(knitr)
opts_chunk$set(message=FALSE, warning=FALSE, eval=TRUE, echo=TRUE)
```

# Data sources

Original data from the Broad Institute's Molecular Signature Database (MSigDB)^[<http://www.broad.mit.edu/gsea/msigdb/index.jsp>], redistributed as separate gmt data files from the MSigDB.

----

_The gene sets contained in the MSigDB are from a wide variety of sources, and relate to a variety of species, mostly human. To facilitate use of the MSigDB in our work, we have created a pure mouse version of the MSigDB by mapping all sets to mouse orthologs. A pure human version is also provided._

**_Prodecure:_** 

_1. The current MSigDB v7.4 gmt files were downloaded from Broad ftp._      
_2. Human Gene symbols were mapped to Mouse Gene symbols, using biomaRt ensembl package (20210815)._      
_3. Each collection was converted to a list in R, and written to a RData file using `save()`._        

----

See the script in `data-raw/` to see how the data frames (tibbles) were created.

# Example usage

There are three data frames (tibbles) this package. The `msigdf.human` data frame has columns for each MSigDB collection divided by sub-collection (like cc, bp and mf for C5). The format of the data is tidy, so each row is a single gene set collection, sub-collection and gene symbol. The `msigdf.mouse` data frame has the same structure for mouse orthologs. The `msigdf.urls` data frame links the name of the gene set to the URL on the Broad's website.

New C5 ontology information into was added into category subcode for easy filtering and consistency

- HPO: Human Phenotype Ontology
- MF: GO Molecular Function ontology
- BP: GO Biological Process ontology
- CC: GO Cellular Component ontology


The data sets in this package have several million rows. The package imports the tibble package so they're displayed nicely.

```{r}
library(tidyverse)
library(msigdf)
```

Take a look:

```{r}
msigdf.human %>% head()
msigdf.mouse %>% head()
msigdf.urls %>% as.data.frame() %>% head()
```

Just get the entries for the [KEGG non-homologous end joining pathway](http://www.genome.jp/kegg/pathway/hsa/hsa03450.html): 

```{r}
msigdf.human %>% 
  filter(geneset=="KEGG_NON_HOMOLOGOUS_END_JOINING")
```

Some software, e.g., [fGSEA](https://github.com/ctlab/fgsea) might require gene sets to be a named list of genes identifiers, where the name of each element in the list is the name of the pathway. This is how the data was originally structured, and we can return to it with `plyr::dlply()`. Here, let's use only the hallmark sets, and after we `dlply` the data into this named list format, get just the first few pathways, and in each of those, just display the first few gene symbols.

```{r}
msigdf.human %>% 
  filter(category_code=="c2") %>% 
  select(geneset, symbol) %>% 
  group_by(geneset) %>% 
  summarize(symbol=list(symbol)) %>% 
  deframe() %>% 
  head() %>% 
  map(head)
```

# Further exploration

For demonstration purposes, create a single object containing both human and mouse data:

```{r}
msigdf <- bind_rows(
  msigdf.human %>% mutate(org="human"),
  msigdf.mouse %>% rename(symbol=mouse.symbol) %>% mutate(org="mouse")
)
```

```{r}
head(msigdf)
tail(msigdf)
```

The number of gene sets in each collection is the same for each organism:

```{r}
msigdf %>%
  group_by(org, category_code) %>%
  summarize(ngenesets=n_distinct(geneset)) %>%
  tidyr::spread(org, ngenesets)
```

But the number of mouse genes in each collection is much greater, due to the one-to-many ortholog mapping.

```{r}
msigdf %>%
  count(org, category_code) %>%
  tidyr::spread(org, n)
```

Look at the first few gene sets just in the 50-geneset hallmark collection. In each gene set, the number of mouse genes is greater than the number of human genes.

```{r}
msigdf %>%
  count(org, category_code, geneset) %>%
  filter(category_code=="h") %>%
  tidyr::spread(org, n)
```

Get the URL for the hallmark set with the fewest number of genes (Notch signaling). Optionally, `%>%` this to `browseURL` to open it up in your browser.

```{r}
msigdf.human %>%
  filter(category_code=="h") %>%
  count(geneset) %>%
  arrange(n) %>%
  head(1) %>%
  inner_join(msigdf.urls, by="geneset") %>%
  pull(url)
```

Just look at the number of genes in each KEGG pathway (sorted descending by the number of genes in that pathway):

```{r}
msigdf.human %>%
  filter(category_code=="c2" & grepl("^KEGG_", geneset)) %>%
  count(geneset) %>% 
  arrange(desc(n))
```


# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
